pkg/syncer/syncer.go

```diff
@@ -224,41 +265,115 @@ func StartSyncer(ctx context.Context, cfg *SyncerConfig, numSyncerThreads int, i
+		upstreamSyncerClusterClient, downstreamDynamicClient, downstreamKubeClient, ddsifForUpstreamSyncer, ddsifForDownstream, downstreamNamespaceController, syncTarget.GetUID(),
+		syncerNamespace, syncerNamespaceInformerFactory, cfg.DNSImage)
 	if err != nil {
 		return err
 	}
 
 	logger.Info("Creating status syncer")
 	statusSyncer, err := status.NewStatusSyncer(logger, logicalcluster.From(syncTarget), cfg.SyncTargetName, syncTargetKey, advancedSchedulingEnabled,
-		upstreamDynamicClusterClient, downstreamDynamicClient, downstreamInformers, syncerInformers, syncTarget.GetUID())
+		upstreamSyncerClusterClient, downstreamDynamicClient, ddsifForUpstreamSyncer, ddsifForDownstream, syncTarget.GetUID())
 	if err != nil {
 		return err
 	}
 
-	upstreamInformers.Start(ctx.Done())
-	downstreamInformers.Start(ctx.Done())
-	kcpInformerFactory.Start(ctx.Done())
-	downstreamInformerFactory.Start(ctx.Done())
+	// Start and sync informer factories
+	var cacheSyncsForAlwaysRequiredGVRs []cache.InformerSynced
+	for _, alwaysRequired := range []string{"secrets", "namespaces"} {
+		gvr := corev1.SchemeGroupVersion.WithResource(alwaysRequired)
+		if informer, err := ddsifForUpstreamSyncer.ForResource(gvr); err != nil {
+			return err
+		} else {
+			cacheSyncsForAlwaysRequiredGVRs = append(cacheSyncsForAlwaysRequiredGVRs, informer.Informer().HasSynced)
+		}
+		if informer, err := ddsifForDownstream.ForResource(gvr); err != nil {
+			return err
+		} else {
+			cacheSyncsForAlwaysRequiredGVRs = append(cacheSyncsForAlwaysRequiredGVRs, informer.Informer().HasSynced)
+		}
+	}
+	ddsifForUpstreamSyncer.Start(ctx.Done())
+	ddsifForUpstreamUpsyncer.Start(ctx.Done())
+	ddsifForDownstream.Start(ctx.Done())
+
+	kcpSyncTargetInformerFactory.Start(ctx.Done())
+	syncerNamespaceInformerFactory.Start(ctx.Done())
 
-	upstreamInformers.WaitForCacheSync(ctx.Done())
-	downstreamInformers.WaitForCacheSync(ctx.Done())
-	kcpInformerFactory.WaitForCacheSync(ctx.Done())
-	downstreamInformerFactory.WaitForCacheSync(ctx.Done())
+	kcpSyncTargetInformerFactory.WaitForCacheSync(ctx.Done())
+	syncerNamespaceInformerFactory.WaitForCacheSync(ctx.Done())
+	cache.WaitForCacheSync(ctx.Done(), cacheSyncsForAlwaysRequiredGVRs...)
 
+	go ddsifForUpstreamSyncer.StartWorker(ctx)
+	go ddsifForUpstreamUpsyncer.StartWorker(ctx)
+	go ddsifForDownstream.StartWorker(ctx)
+
+	// Start static controllers
 	go apiImporter.Start(klog.NewContext(ctx, logger.WithValues("resources", resources)), importPollInterval)
-	go syncerInformers.Start(ctx, 1)
+	go syncTargetGVRSource.Start(ctx, 1)
 	go specSyncer.Start(ctx, numSyncerThreads)
 	go statusSyncer.Start(ctx, numSyncerThreads)
+
 	go downstreamNamespaceController.Start(ctx, numSyncerThreads)
 
+	// Create and start GVR-specific controllers through controller managers
+	upstreamSyncerControllerManager := controllermanager.NewControllerManager(ctx,
```

Comment:
this ordering is dangerous. Please always do this:
1. create informers
2. create controllers
3. start informers
4. wait for informers
5. start controllers